man/man1/gflsof.1
'\" t
.\"     Title: gflsof
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 27 Sep 2016
.\"    Manual: Gfarm
.\"    Source: Gfarm
.\"  Language: English
.\"
.TH "GFLSOF" "1" "27 Sep 2016" "Gfarm" "Gfarm"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gflsof \- list currently opened Gfarm files
.SH "SYNOPSIS"
.HP \w'\fBgflsof\fR\ 'u
\fBgflsof\fR [\-AW] [\-P\ \fIpath\fR] [\-D\ \fIdomainname\-of\-gfsd\fR] [\-U\ \fIdomainname\-of\-user\-host\fR] [\-u\ \fIusername\fR]
.HP \w'\fBgflsof\fR\ 'u
\fBgflsof\fR [\-AW] [\-P\ \fIpath\fR] [\-D\ \fIdomainname\-of\-gfsd\fR] [\-U\ \fIdomainname\-of\-user\-host\fR] [\-u\ \fIusername\fR] \-C
.HP \w'\fBgflsof\fR\ 'u
\fBgflsof\fR [\-AW] [\-P\ \fIpath\fR] [\-D\ \fIdomainname\-of\-gfsd\fR] [\-U\ \fIdomainname\-of\-user\-host\fR] [\-u\ \fIusername\fR] \-G
.SH "DESCRIPTION"
.PP
Displays information about each Gfarm file descriptors, if neither
\fB\-C\fR
nor
\fB\-G\fR
is specified\&. The USER field shows username, the GPID field shows Gfarm process ID, the FD field shows Gfarm file descriptor number, if the character just after the descriptor number is \'r\', it means read\-only open, if the character is \'w\', it means write\-only open, and if the character is \'u\', it means read\-write open, the TYPE fields shows the type of the inode (REG is a file, DIR is a directory and LNK is a symbolic link), the INODE field shows the inode number, the GEN field shows the generation number of the inode, the SIZE/OFF shows the seek\-offset value if the inode is a directory, or the file size if the inode is a file, the CLIENT field shows hostname and port number of the client which is connecting to the gfmd, the GFSD field shows hostname and port number of the gfsd which is connecting to the gfmd\&.
.PP
If
\fB\-C\fR
option is specified, this command shows only hostnames of clients which are opening descriptors\&.
.PP
If
\fB\-G\fR
option is specified, this command shows only hostnames of gfsd which are opening file replicas\&.
.SH "OPTIONS"
.PP
\fB\-A\fR
.RS 4
Display information about all users\&. Only gfarmadm group is allowed to use this option\&.
.RE
.PP
\fB\-W\fR
.RS 4
Limit subject descriptors to which are either write\-only opened or read\-write opened\&.
.RE
.PP
\fB\-P\fR \fIpath\fR
.RS 4
Specify a pathname or a Gfarm URL to identify metadata server which will be accessed by this command\&. If not specified, "\&." is used as the default value\&.
.RE
.PP
\fB\-D\fR \fIdomainname\-of\-gfsd\fR
.RS 4
Limit subject descriptors to which gfsd is currently opening a replica and the gfsd belongs to the specified domain\&.
.RE
.PP
\fB\-U\fR \fIdomainname\-of\-user\-host\fR
.RS 4
Limit subject descriptors to which client host belongs to the specified domain\&.
.RE
.PP
\fB\-u\fR \fIusername\fR
.RS 4
Limit subject descriptors to which is opened by the specified user\&. Only gfarmadm group is allowed to use this option\&.
.RE
.PP
\fB\-?\fR
.RS 4
Displays a list of command options\&.
.RE
.SH "EXAMPLES"
.PP
\fBExample\ \&1.\ \&list file descriptors which are opened by myself\fR
.PP

.sp
.if n \{\
.RS 4
.\}
.nf
gflsof
.fi
.if n \{\
.RE
.\}
.sp

.PP
\fBExample\ \&2.\ \&list all users' file descriptors\fR
.PP

.sp
.if n \{\
.RS 4
.\}
.nf
gflsof \-A
.fi
.if n \{\
.RE
.\}
.sp

.PP
\fBExample\ \&3.\ \&list filesystem nodes which are currently opening file replicas in writable mode\fR
.PP

.sp
.if n \{\
.RS 4
.\}
.nf
gflsof \-AWG
.fi
.if n \{\
.RE
.\}
.sp

.PP
\fBExample\ \&4.\ \&list descriptors that corresponding gfsd processes are opening file replicas in writable mode\fR
.PP

.sp
.if n \{\
.RS 4
.\}
.nf
gflsof \-AW \-D \'\'
.fi
.if n \{\
.RE
.\}
.sp

.PP
\fBExample\ \&5.\ \&list client hosts which belong to example.org domain and user john is opening descriptors on the hosts\fR
.PP

.sp
.if n \{\
.RS 4
.\}
.nf
gflsof \-U example\&.org \-u john \-C
.fi
.if n \{\
.RE
.\}
.sp




ChangeLog
2007-03-31  Noriyuki Soda  <soda@sra.co.jp>

	* RELNOTES
	avoid an expression which may be misinterpreted as a security problem.

2007-03-31  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* RELNOTES
	correct description

2007-03-30  Tohru Sotoyama <sotoyama@sra.co.jp>

	* package/solaris/build
	change globus root directory

2007-03-30  Noriyuki Soda  <soda@sra.co.jp>

	* ChangeLog LICENSE RELNOTES
	merge changes in the CVS main trunk until gfarm-1.4.1 to the GNEO
	branch
	
	* ChangeLog
	Catched up with the current ChangeLog.

	* ChangeLog
	recover missed log messages in previous commit
	
	* doc/html/en/ref/man5/gfarm.conf.5.html, doc/html/ja/ref/man5/gfarm.conf.5.html man/ja/man5/gfarm.conf.5 man/man5/gfarm.conf.5:
	regen

	* doc/Gfarm-FAQ.en, doc/Gfarm-FAQ.ja, doc/docbook/en/ref/man5/gfarm.conf.5.docbook, doc/docbook/ja/ref/man5/gfarm.conf.5.docbook:
	merge the following changes in the SourceForge Subversion repository:

	Revision: 3674
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3674&view=rev
	Author:   sotoyama
	Date:     2007-03-29 02:34:39 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814]

	Revision: 3676
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3676&view=rev
	Author:   sotoyama
	Date:     2007-03-29 04:46:11 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814] to `auth' statement

	Revision: 3682
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3682&view=rev
	Author:   n-soda
	Date:     2007-03-29 11:03:31 -0700 (Thu, 29 Mar 2007)
	Log Message:
	swap section order of limitation and example of the "auth" directive,
	and add a paragraph to describe the example.

2007-03-30  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES:
	proofreading

2007-03-29  Noriyuki Soda  <soda@sra.co.jp>

	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3681
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3681&view=rev
	Author:   n-soda
	Date:     2007-03-29 07:37:36 -0700 (Thu, 29 Mar 2007)
	Log Message:
	report error status of close(2) in gfs_server_open_local(),
	problem reported by Markus Elfring as:
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687850&group_id=172718&atid=862742

	Also, treat an error at sendmsg(2) as just an error instead of
	fatal one.

	* bench/fsysbench/fsysbench.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3679
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3679&view=rev
	Author:   n-soda
	Date:     2007-03-29 05:13:05 -0700 (Thu, 29 Mar 2007)
	Log Message:
	timer_interrupt_handler() shouldn't call printf(3), but write(2),
	because the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* ftool/gfarm_agent/gfarm_agent.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3678
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3678&view=rev
	Author:   n-soda
	Date:     2007-03-29 04:59:54 -0700 (Thu, 29 Mar 2007)
	Log Message:
	sigterm_handler() shouldn't call exit(3), but _exit(2), because
	the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3675
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3675&view=rev
	Author:   n-soda
	Date:     2007-03-29 03:30:02 -0700 (Thu, 29 Mar 2007)
	Log Message:
	do not call exit(3) but _exit(2) in accepting_sigterm_handler(),
	because the former is not async-signal-safe.
	problem reported by Markus Elfring in
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687842&group_id=172718&atid=862742

2007-03-29  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES, LICENSE:
	update release notes and the copyright display for the version 1.4.1

2007-03-28  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* ChangeLog:
	Catched up with the current ChangeLog.

2007-03-28  Noriyuki Soda  <soda@sra.co.jp>

	* include/gfarm/gfarm_misc.h, package/freebsd/Makefile,	package/pkgsrc/Makefile, package/redhat/gfarm.spec, package/solaris/build:
	increase version number to 1.4.1

	* include/gfarm/gfarm_error.h, lib/libgfarm/gfarm/error.c, lib/libgfarm/gfarm/gfs_client.c:
	merge the following change from the SourceForge Subversion repository:

	Revision: 3673
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3673&view=rev
	Author:   n-soda
	Date:     2007-03-28 07:11:37 -0700 (Wed, 28 Mar 2007)
	Log Message:
	make gfsd reconnection feature work against more network errors.
	this fixes a problem reported by Yamamoto Naotaka-san
	in [gfarm-developers:03456] which was caused by ECONNRESET.

	introduce ENETDOWN, EHOSTDOWN and ENETRESET as gfarm native errors
	for this purpose. (although ENETRESET will be nearly never used)

	* gftool/gneo/gneo.c:
	just a portability fix

	* regress/gfs_hook/lib/fsystest/index.FreeBSD.expected, regress/gfs_hook/lib/fsystest/index.Linux.expected, regress/gfs_hook/lib/fsystest/index.NetBSD.expected, regress/gfs_hook/lib/fsystest/index.SunOS.expected:
	results are different on the GNEO branch with gfs_hook


gftool/gflsof/gflsof.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>

#include <gfarm/gfarm.h>

#include "hash.h"

#include "gfm_proto.h"
#include "gfm_client.h"
#include "lookup.h"
#include "gfarm_path.h"

#define HOSTNAME_HASHTAB_SIZE	3079	/* prime number */
#define BUF_LEN			1024

enum operation_mode {
	OP_CLIENT_LIST = 'C',
	OP_GFSD_LIST = 'G',
	OP_DEFAULT = '\0'
};

static const char OPT_USER_ALL[] = "";
static const char *program_name = "gflsof";

static struct gfarm_hash_table *hostname_hashtab;

static void
hostname_hashtab_init(void)
{
	hostname_hashtab =
	    gfarm_hash_table_alloc(HOSTNAME_HASHTAB_SIZE,
		gfarm_hash_casefold, gfarm_hash_key_equal_casefold);
	if (hostname_hashtab == NULL) {
		fprintf(stderr, "%s: no memory for %d entry hash table\n",
		    program_name, HOSTNAME_HASHTAB_SIZE);
		exit(EXIT_FAILURE);
	}
}

static int
hostname_hashtab_enter(const char *str)
{
	int created;

	if (gfarm_hash_enter(hostname_hashtab, str, strlen(str) + 1, 1,
	    &created) == NULL) {
		fprintf(stderr, "%s: no memory to record `%s'\n",
		    program_name, str);
		exit(EXIT_FAILURE);
	}
	return (created);
}

static void
print_client_list(int nfds, struct gfarm_process_fd_info *fd_info)
{
	int i;
	struct gfarm_process_fd_info *fdi;

	hostname_hashtab_init();

	for (i = 0; i < nfds; i++) {
		fdi = &fd_info[i];
		if (hostname_hashtab_enter(fdi->fd_client_host))
			printf("%s\n", fdi->fd_client_host);
	}
}

static void
print_gfsd_list(int nfds, struct gfarm_process_fd_info *fd_info)
{
	int i;
	struct gfarm_process_fd_info *fdi;

	hostname_hashtab_init();

	for (i = 0; i < nfds; i++) {
		fdi = &fd_info[i];
		if (hostname_hashtab_enter(fdi->fd_gfsd_host))
			printf("%s\n", fdi->fd_gfsd_host);
	}
}

static void
set_lenmax(int *lenmax, const char *s)
{
	int len = strlen(s);

	if (*lenmax < len)
		*lenmax = len;
}

static void
print_fds(int nfds, struct gfarm_process_fd_info *fd_info)
{
	int i;
	struct gfarm_process_fd_info *fdi;
	int user_width = 4; /* == strlen("USER"), and so on... */
	int gpid_width = 4;
	int fd_width = 2;
	int inode_width = 5;
	int gen_width = 3;
	int off_width = 8; /* == strlen("SIZE/OFF") */
	int client_width = 6;
	int gfsd_width = 4;
	char buf[BUF_LEN];

	for (i = 0; i < nfds; i++) {
		fdi = &fd_info[i];
		snprintf(buf, sizeof buf, "%s", fdi->fd_user);
		set_lenmax(&user_width, buf);
		snprintf(buf, sizeof buf, "%llu", (long long)fdi->fd_pid);
		set_lenmax(&gpid_width, buf);
		snprintf(buf, sizeof buf, "%u", (int)fdi->fd_fd);
		set_lenmax(&fd_width, buf);
		snprintf(buf, sizeof buf, "%llu", (long long)fdi->fd_ino);
		set_lenmax(&inode_width, buf);
		snprintf(buf, sizeof buf, "%llu", (long long)fdi->fd_gen);
		set_lenmax(&gen_width, buf);
		snprintf(buf, sizeof buf, "%llu", (long long)fdi->fd_off);
		set_lenmax(&off_width, buf);
		snprintf(buf, sizeof buf, "%s:%u",
		    fdi->fd_client_host, (int)fdi->fd_client_port);
		set_lenmax(&client_width, buf);
#if 0
		snprintf(buf, sizeof buf, "%s:%u:%u",
		    fdi->fd_gfsd_host, (int)fdi->fd_gfsd_port,
		    (int)fdi->fd_gfsd_peer_port);
#else
		snprintf(buf, sizeof buf, "%s:%u",
		    fdi->fd_gfsd_host, (int)fdi->fd_gfsd_peer_port);
#endif
		set_lenmax(&gfsd_width, buf);
	}

	printf("%*s %*s %*s %4s %*s %*s %*s %*s %*s\n",
	    -user_width, "USER",
	    gpid_width, "GPID",
	    fd_width, "FD",
	    "TYPE",
	    inode_width, "INODE",
	    gen_width, "GEN",
	    off_width, "SIZE/OFF",
	    -client_width, "CLIENT",
	    -gfsd_width, "GFSD");
	for (i = 0; i < nfds; i++) {
		fdi = &fd_info[i];
		printf("%*s %*llu %*u",
		       -user_width, fdi->fd_user,
		       gpid_width, (long long)fdi->fd_pid,
		       fd_width, (int)fdi->fd_fd);
		switch (fdi->fd_open_flags & GFARM_FILE_ACCMODE) {
		case GFARM_FILE_RDONLY:
			putchar('r'); break;
		case GFARM_FILE_WRONLY:
			putchar('w'); break;
		case GFARM_FILE_RDWR:
			putchar('u'); break;
		default: /* LOOKUP */
			putchar(' '); break;
		}
		switch (fdi->fd_mode & GFARM_S_IFMT) {
		case GFARM_S_IFDIR:
			printf(" DIR"); break;
		case GFARM_S_IFREG:
			printf(" REG"); break;
		case GFARM_S_IFLNK:
			printf(" LNK"); break;
		default:
			printf(" %03o", (fdi->fd_mode & GFARM_S_IFMT) >> 12);
			break;
		}
		printf(" %*llu %*llu %*llu",
		    inode_width, (long long)fdi->fd_ino,
		    gen_width, (long long)fdi->fd_gen,
		    off_width, (long long)fdi->fd_off);

		if (fdi->fd_client_host[0] != '\0')
			snprintf(buf, sizeof buf, "%s:%u",
			    fdi->fd_client_host, (int)fdi->fd_client_port);
		else
			snprintf(buf, sizeof buf, "-");
		printf(" %*s", -client_width, buf);

		if (fdi->fd_gfsd_host[0] != '\0')
#if 0
			snprintf(buf, sizeof buf, "%s:%u:%u",
			    fdi->fd_gfsd_host, (int)fdi->fd_gfsd_port,
			    (int)fdi->fd_gfsd_peer_port);
#else
			snprintf(buf, sizeof buf, "%s:%u",
			    fdi->fd_gfsd_host, (int)fdi->fd_gfsd_peer_port);
#endif
		else
			snprintf(buf, sizeof buf, "-");
		printf(" %*s\n", -gfsd_width, buf);
	}
}

static void
usage(void)
{
	fprintf(stderr,
	    "Usage: %s [-ACGW] [-D <gfsd_domain>] [-U <domain_of_user_host>] "
	    "[-P <path>] [-u <user>]\n",
	    program_name);
	exit(2);
}

static void
check_and_set_op(enum operation_mode *op_modep, enum operation_mode op)
{
	if (*op_modep != OP_DEFAULT) {
		fprintf(stderr, "%s: -%c option conflicts with -%c\n",
		    program_name, op, *op_modep);
		usage();
		/*NOTREACHED*/
	}
	*op_modep = op;
}

static void
check_and_set_user(const char **userp, const char *arg)
{
	if (*userp != NULL) {
		if (strcmp(*userp, OPT_USER_ALL) == 0)
			fprintf(stderr, "%s: -u %s option conflicts "
			    "with -A\n", program_name, arg);
		else
			fprintf(stderr, "%s: -u %s option conflicts "
			    "with -u %s\n", program_name, arg, *userp);
		usage();
		/*NOTREACHED*/
	}
	*userp = arg;
}

int
main(int argc, char **argv)
{
	gfarm_error_t e;
	int exit_code = EXIT_SUCCESS;
	struct gfm_connection *gfm_server = NULL;
	enum operation_mode op_mode = OP_DEFAULT;
	char *opt_gfsd_domain = "";
	char *opt_user_host_domain = "";
	const char *opt_user = NULL;
	gfarm_uint64_t opt_flags = 0;

	static const char OPT_PATH_DEFAULT[] = ".";
	const char *opt_path = OPT_PATH_DEFAULT;
	char *realpath = NULL;

	int c, nfds;
	struct gfarm_process_fd_info *fd_info;

	if (argc > 0)
		program_name = basename(argv[0]);

	e = gfarm_initialize(&argc, &argv);
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n", program_name,
		    gfarm_error_string(e));
		exit(EXIT_FAILURE);
	}

	while ((c = getopt(argc, argv, "CGAD:P:U:Wu:?")) != -1) {
		switch (c) {
		case OP_CLIENT_LIST:
			check_and_set_op(&op_mode, c);
			opt_flags |=
			    GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_CLIENT_DETACH;
			break;
		case OP_GFSD_LIST:
			check_and_set_op(&op_mode, c);
			opt_flags |=
			    GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_GFSD_DETACH |
			    (GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_INODE_ALLMASK ^
			     GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_INODE_REG);
			break;
		case 'A':
			check_and_set_user(&opt_user, OPT_USER_ALL);
			break;
		case 'D':
			opt_gfsd_domain = optarg;
			opt_flags |=
			    GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_GFSD_DETACH |
			    (GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_INODE_ALLMASK ^
			     GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_INODE_REG);
			break;
		case 'P':
			opt_path = optarg;
			break;
		case 'U':
			opt_user_host_domain = optarg;
			opt_flags |=
			    GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_CLIENT_DETACH;
			break;
		case 'W':
			opt_flags |=
			    GFM_PROTO_PROCESS_FD_FLAG_EXCLUDE_WRITE_NO_OPEN;
			break;
		case 'u':
			check_and_set_user(&opt_user, optarg);
			break;
		case '?':
		default:
			usage();
			/*NOTREACHED*/
		}
	}
	argc -= optind;
	argv += optind;
	if (argc > 0) {
		fprintf(stderr, "%s: extra operand `%s'\n",
		    program_name, argv[0]);
		usage();
		/*NOTREACHED*/
	}

	if (gfarm_realpath_by_gfarm2fs(opt_path, &realpath)
	    == GFARM_ERR_NO_ERROR)
		opt_path = realpath;
	if ((e = gfm_client_connection_and_process_acquire_by_path(
	    opt_path, &gfm_server)) != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: metadata server for \"%s\": %s\n",
		    program_name, opt_path, gfarm_error_string(e));
		exit(EXIT_FAILURE);
	}
	free(realpath);

	if (opt_user == NULL)
		opt_user = gfm_client_username(gfm_server);

	e = gfm_client_process_fd_info(gfm_server,
	    opt_gfsd_domain, opt_user_host_domain, opt_user, opt_flags,
	    &nfds, &fd_info);
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n",
		    program_name, gfarm_error_string(e));
		exit_code = EXIT_FAILURE;
	} else {
		switch (op_mode) {
		case OP_CLIENT_LIST:
			print_client_list(nfds, fd_info);
			break;
		case OP_GFSD_LIST:
			print_gfsd_list(nfds, fd_info);
			break;
		default:
			print_fds(nfds, fd_info);
			break;
		}
		gfarm_process_fd_info_free(nfds, fd_info);
	}

	e = gfarm_terminate();
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n", program_name,
		    gfarm_error_string(e));
		exit(EXIT_FAILURE);
	}
	return (exit_code);
}
